corehq/apps/locations/sql_templates/get_location_fixture_ids_2.sql

```diff
@@ -0,0 +1,333 @@
+
+        "expand_from" columns:
+        - loc_id: location id, null for include_without_expanding or
+          expand_from_root.
+        - depth: expand to depth. Negative values in this column have
+          special meanings. See output examples below.
+
+         loc_id | depth
+        --------|-------
+         NULL   |  3     -- include all locations with depth <= 3
+         1      |  4     -- include all descendents of location 1 to depth 4
+         10     | -1     -- include location 10 (but do not expand)
+         100    | -2     -- include all descendents of location 100, unlimited depth
+         11     | -3     -- location 11 and its descendants are included based on
+                            include_only types
+        */
+
+        WITH RECURSIVE cte AS (
+            -- get include_without_expanding depth
+            SELECT
+                NULL AS "parent_id",
+                NULL AS "expand_from_type",
+                NULL AS "loc_id",
+                "expand_to"."expand_to_depth" AS "depth"
+            FROM "expand_to"
+            WHERE "expand_to"."expand_to_type" = -1
+
+            UNION ALL
+
+            SELECT
+                loc."parent_id",
+                CASE
+                    WHEN (
+                        -- if expand_from is set and not the current location type
+                        -- it will be one of this location's ancestors
+                        loc_type."expand_from" IS NOT NULL
+                        AND loc_type."expand_from_root" = FALSE
+                        AND loc_type."expand_from" <> loc."location_type_id"
+                        AND NOT EXISTS (
+                            -- might be wrong to ignore loc_type.expand_from
+                            -- when include_only types exist
+                            SELECT 1
+                            FROM "locations_locationtype_include_only"
+                            WHERE "from_locationtype_id" = loc."location_type_id"
+                        )
+                    ) AND NOT case_sync_restriction THEN loc_type."expand_from"
+                    -- otherwise it will be null for this and all ancestors
+                    ELSE NULL
+                END AS "expand_from_type",
+                CASE
+                    -- expand_from_root -> no path
+                    WHEN loc_type."expand_from_root" = TRUE AND NOT case_sync_restriction THEN NULL
+                    -- else first path element
+                    ELSE loc."id"
+                END AS "loc_id",
+                CASE
+                    WHEN case_sync_restriction THEN (
+                        CASE WHEN loc_type."restrict_cases_to_id" IS NOT NULL THEN (
+                            SELECT "expand_to_depth"
+                            FROM "expand_to"
+                            WHERE "expand_to_type" = loc_type."restrict_cases_to_id"
+                        ) ELSE -2
+                        END
+                    )
```

Comment:
Fair enough, although that pushes me closer to thinking this function is being abused. Maybe the `case_sync_restriction` should be implemented as a separate function since it appears to be a quite different use case?

Reply:
That seems like a win if you can structure it that way.